Enable anisotropic filtering for 3D materials by default on desktop platforms

Calinou opened the pull request:
This results in better appearance for textures viewed from oblique angles. Many other engines now enable it by default. On today's desktop platforms, the performance impact is minimal if other effects are already taxing the GPU.

The anisotropic filtering quality can be configured as usual in the Project Settings. Setting it to 1 should disable it even if individual materials have it enabled, so this is easy to revert if you don't want it.

This remains disabled by default on platforms with mobile (TBDR) GPUs for performance reasons, as texture reads are often the bottleneck there.

I see screenshots of Godot projects not using anisotropic filtering too often – let's change that! :slightly_smiling_face: 

This closes https://github.com/godotengine/godot-proposals/issues/3134.

## Preview

*I used [this Sponza demo](https://github.com/Calinou/godot-sponza-4) for the screenshots. Click to view at full size, look at the floor.*

### Before

![aniso_off](https://user-images.githubusercontent.com/180032/79124530-cc0d1100-7d9c-11ea-8d5a-61b17bd9e1ee.png)
	
### After

*Note: the screenshot below uses 4× anisotropic filtering, not 8× which is the new default on desktop platforms.*

![aniso_on](https://user-images.githubusercontent.com/180032/79123301-248edf00-7d9a-11ea-8294-c34871868f7a.png)

Logharaa commented:
Why is it not merged?

Calinou commented:
@Logharaa reduz said on IRC he'd prefer this to be implemented in a more fine-grained fashion.

Logharaa commented:
@Calinou Oh ok thanks for the info!

reduz commented:
Mobile cant deal with this as well, and it needs to be tested on lower end devices too, so lets wait a bit on this one

Calinou commented:
I amended the pull request to use 8× anisotropic filtering desktop and 1× on mobile (which effectively disables it there).

reduz commented:
This needs testing on low end desktop devices to ensure it makes a difference in performance or not.

fire commented:
@Calinou Do you think we can cover enough devices to turn this on by default?

fire commented:
I salvaged the PR, but I'll let it simmer in my branch of Godot Engine for a bit until I submit it.

https://github.com/V-Sekai/godot/tree/vsk-default-enable-anisotropic-filtering-4.4

Calinou committed:
Enable anisotropic filtering for 3D materials by default on desktop

This results in better appearance for textures viewed from oblique
angles. Many other engines now enable it by default.
On today's desktop platforms, the performance impact is minimal if
other effects are already taxing the GPU.

The default value chosen (8x) is a good balance between performance
and quality. Depending on the scene, you may be able to use 16x
anisotropic filtering with little to no cost in terms of frame time.

The anisotropic filtering quality can be configured as usual in the
Project Settings. Setting it to 1x will disable it even if individual
materials have it enabled, so this is easy to revert if you don't
want it.

This remains disabled by default on platforms with mobile (TBDR) GPUs
for performance reasons, as texture reads are often the bottleneck
there.

Calinou commented:
Rebased and tested again (using @fire's work as a base), it works as expected.

I've also modified the glTF importer to handle this for material import (remember to reimport 3D scenes for this change to apply, e.g. by removing the `.godot/` folder).

fire commented on servers/rendering_server.cpp:
```diff
@@ -3603,7 +3603,10 @@ void RenderingServer::init() {
-	GLOBAL_DEF_RST(PropertyInfo(Variant::INT, "rendering/textures/default_filters/anisotropic_filtering_level", PROPERTY_HINT_ENUM, String::utf8("Disabled (Fastest),2× (Faster),4× (Fast),8× (Average),16× (Slow)")), 2);
+	GLOBAL_DEF_RST(PropertyInfo(Variant::INT, "rendering/textures/default_filters/anisotropic_filtering_level", PROPERTY_HINT_ENUM, String::utf8("Disabled (Fastest),2× (Faster),4× (Fast),8× (Average),16× (Slow)")), 3); // 8x anisotrophic filtering.
+
+	// Disable anisotropic filtering on mobile by default as it's relatively expensive.
```
On mobile anti-aliasing 4x is free, I don't know about aniso. Do you know?

Repiteo reviewed (approved):
Codestyle checks out. Kinda surprised that anisotropic filtering *wasn't* already the default; always assumed it was a pretty free feature (maybe not for **really** low-end devices?)

clayjohn commented:
I agree with reduz, this needs confirmation that it won't harm low end devices. I'm confident that it is a good change for most dedicated GPUs, but I suspect that many integrated GPUs will take a performance hit from this. Since this imposes a new default on all users, we need to be careful and validate these changes before we go ahead.

Calinou commented:
I did some tests of https://github.com/Calinou/godot-sponza-4 on my laptop (on AC in balanced mode):

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900H
- **GPU:** Intel Iris Xe / NVIDIA GeForce RTX 4060 Laptop GPU
- **RAM:** 48 GB (16 + 32 GB DDR4-3200 C22)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Windows 11 23H2
</details>

Type | `master` | This PR
-|-|-
Intel, Mobile, 1152x648 | 296 FPS (3.38 mspf) | 283 FPS (3.54 mspf)
Intel, Mobile, 2560x1600 | 90 FPS (11.11 mspf) | 88 FPS (11.36 mspf)
Intel, Compatibility, 1152x648 | 490 FPS (2.04 mspf) | 492 FPS (2.03 mspf)
Intel, Compatibility, 2560x1600 | 144 FPS (6.94 mspf) | 140 FPS (7.14 mspf)
NVIDIA, Mobile, 1152x648 | 1510 FPS (0.66 mspf) | 1487 FPS (0.67 mspf)
NVIDIA, Mobile, 2560x1600 | 611 FPS (1.64 mspf) | 599 FPS (1.67 mspf)
NVIDIA, Compatibility, 1152x648 | 1695 FPS (0.59 mspf) | 1727 FPS (0.58 mspf)
NVIDIA, Compatibility, 2560x1600 | 798 FPS (1.25 mspf) | 797 FPS (1.25 mspf)

The performance difference is very small on average, especially compared to the significant gain in quality. Limiting ourselves to 8x probably did the heavy lifting here, since it's basically twice as cheap as 16x with only a slight loss in visual quality. (Console versions of games often limit themselves to 8x AF for a similar reason, since CPU and GPU memory is shared.)

Note that this is a scene with no post-processing effects applied, so scenes with post-processing will spend a greater time in areas other than anisotropic texture sampling (therefore reducing the % impact of anisotropic filtering).

Zireael07 commented:
What about Android devices? They're the ones which are really low-end for Godot...

Calinou commented:
> What about Android devices? They're the ones which are really low-end for Godot...

I've tested 8x AF on my Samsung Galaxy Z Fold4 and couldn't spot a performance difference on its main screen (it always ran at a locked 120 FPS, also on Truck Town).

However, this PR doesn't enable anisotropic filtering by default on mobile (since the `.mobile` feature tag override is used to set it to 1x, which is the same as disabled). We could revisit this in a future PR - I expect 2x AF to perform well on all mobile devices nowadays, maybe even 4x.

clayjohn commented:
> However, this PR doesn't enable anisotropic filtering by default on mobile (since the .mobile feature tag override is used to set it to 1x, which is the same as disabled). We could revisit this in a future PR - I expect 2x AF to perform well on all mobile devices nowadays, maybe even 4x.

The performance impact of anisotropic textures doesn't work like that. There is no inherent performance penalty from using it (except for on very old devices), the performance cost comes from the additional bandwidth needed to sample it. Mobile devices are often bottlenecked by reading from textures, so multiplying the cost of texture reads is prohibitive. Desktop devices are almost never bottlenecked by texture reads, so its a free quality improvement. Remember, taking more time for something that isn't the bottleneck is essentially free (until you optimize your existing bottleneck). 

Low end integrated devices struggle more with texture reads than dedicated GPUs do. So we need to test there to assess the realistic impact. Testing on a high end device or a dedicated GPU won't give you any useful information unfortunately

lyuma reviewed (approved):
This seems good to bring Godot up to industry standard texture quality on PC platforms.

clayjohn commented:
I was doing some performance profiling on macOS last week of the TPS demo and I found that anisotropic textures were a significant cause of high bandwidth in that demo. TBDR GPUs tend to get bandwidth limited fast, so increasing bandwidth is a big concern. It would be nice if this could be scoped to non-TBDR GPUs. I'm not sure if there is an easy way for us to do that.

Unfortunately this means that this PR will likely hurt performance on any ARM-based laptops as well as all mobile devices. It might be a tradeoff worth making, but I'm not sure if it is

fire commented:
We discussed a long time ago where system where known incompatible systems degrade to a more performant setting. Unknown systems assume to work better.

Calinou commented:
> TBDR GPUs tend to get bandwidth limited fast, so increasing bandwidth is a big concern. It would be nice if this could be scoped to non-TBDR GPUs. I'm not sure if there is an easy way for us to do that.

If we had a way to use multiple feature tags at once in the Project Settings, we could use `.macos.arm` as an override to disable anisotropic filtering like we currently do for `.mobile`.

fire commented:
If your platform is a nintendo switch your platform provider probably wants a hook

v1993 commented:
From my perspective this should be a pretty easy merge in its current state, as this gives a noticeable quality boost on default settings and primarily *enables* proper control over anisotropic filtering (via the existing project setting), meaning that the feature isn't forced on anyone and can be easily opted out of if it turns out to cause problems for a project. I don't think there's really a need for more override options, at least for the initial merge - on desktop anisotropic filtering is traditionally exposed as a setting to end users anyways, which those with a TBDR GPU could turn down manually if they get bottlenecked by it.

lyuma commented:
We discussed this in the asset pipeline meeting. Anisotropic filtering is industry standard and standard on desktop platforms.

Because this PR combines two areas of the engine, it is not possible to get consensus with this change as written.

We decided that the best course of action is to wait for Godot 4.8, and early in the merge cycle, change the default for *all importers*, to use ANISOTROPIC filter mode: This is the artist intent.

We will then file an issue for the rendering team to test and make sure that all platforms have good defaults with regards to anisotropy, or to add the necessary performance controls to allow textures imported as anisotropic filtering to run without anisotropy or not otherwise cause performance degradation on mobile architectures.

This way, we have the whole release cycle of 4.8 to work out any performance issues related to this change in the importers.

Then, this PR can be repurposed to remove the changes to default settings on the material resources, and only focus on rendering server and project settings changes.

clayjohn commented:
> We will then file an issue for the rendering team to test and make sure that all platforms have good defaults with regards to anisotropy, or to add the necessary performance controls to allow textures imported as anisotropic filtering to run without anisotropy or not otherwise cause performance degradation on mobile architectures.

Anisotropy isn't an import setting, its a sampler setting which gets set by the shader/material. No changes to the importer are needed. IMO intentionally breaking everyones workflows and harming performance for all games and then telling the rendering team to fix it is not a plan that should be followed. 

This is an area where having the ability to configure defaults at project creation time is important. Users should be able to select whether they are targeting Desktop or Mobile devices. If they select Desktop in the project configuration, _then_ you can do things like this and enable anisotropy by default. Otherwise, it should not be forced on anyone. Anisotropy is only cheap on desktop devices with dedicated GPUs, everywhere else pays a huge cost.

Calinou commented:
> This is an area where having the ability to configure defaults at project creation time is important. Users should be able to select whether they are targeting Desktop or Mobile devices. If they select Desktop in the project configuration, then you can do things like this and enable anisotropy by default. Otherwise, it should not be forced on anyone. Anisotropy is only cheap on desktop devices with dedicated GPUs, everywhere else pays a huge cost.

Could we have a `dedicated_gpu` feature tag, similar to what I proposed above? This would allow for defaults to be set according to the GPU type automatically, without requiring users to go through an initial setup dialog.

I agree we'll need some kind of setup dialog in the long run (e.g. to select [settings optimized for a pixel art style](https://github.com/godotengine/godot-proposals/issues/5658), or on the contrary, high-end 3D). That said, I have a few concerns related to it, e.g. when users want to select another preset after creating their project. There's surely a solution that works both before and after creating a project, e.g. by letting you switch presets in the Project Settings dialog, but it might be more complex to implement this way.

I might look into implementing setting presets after 4.7. I'll open a proposal soon with how I envision the feature.

**Edit:** Proposal opened:

- https://github.com/godotengine/godot-proposals/issues/15320
